src/main/java/gregtech/api/util/OverlayedItemHandler.java

```diff
@@ -0,0 +1,119 @@
+package gregtech.api.util;
+
+import gregtech.api.recipes.KeySharedStack;
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.items.IItemHandler;
+
+import javax.annotation.Nonnull;
+
+public class OverlayedItemHandler {
+    private final OverlayedItemHandlerSlot[] originalSlots;
+    private final OverlayedItemHandlerSlot[] slots;
+    private final IItemHandler overlayedHandler;
+    private ItemStackKey cachedStackKey;
+    private ItemStack cachedStack;
+
+    public OverlayedItemHandler(IItemHandler toOverlay) {
+        this.slots = new OverlayedItemHandlerSlot[toOverlay.getSlots()];
+        this.originalSlots = new OverlayedItemHandlerSlot[toOverlay.getSlots()];
+        this.overlayedHandler = toOverlay;
+    }
+
+    /**
+     * Resets the {slots} array to the state when the handler was
+     * first mirrored
+     */
+
+    public void reset() {
+        for (int i = 0; i < this.originalSlots.length; i++) {
+            if (this.originalSlots[i] != null) {
+                this.slots[i] = this.originalSlots[i].copy();
+            }
+        }
+    }
+
+    public int getSlots() {
+        return overlayedHandler.getSlots();
+    }
+
+    private void initSlot(int slot) {
+        if (this.originalSlots[slot] == null) {
+            ItemStack stackToMirror = overlayedHandler.getStackInSlot(slot);
+            this.originalSlots[slot] = new OverlayedItemHandlerSlot(stackToMirror);
+            this.slots[slot] = new OverlayedItemHandlerSlot(stackToMirror);
+        }
+    }
+
+    public int insertItemStackKey(int slot, @Nonnull ItemStackKey toInsert, int amount) {
+        initSlot(slot);
+
+        ItemStack stack = null;
+        if (toInsert == cachedStackKey) {
+            if (cachedStack != null) {
+                stack = cachedStack;
+            }
+        }
+
+        ItemStack remainder;
+        if (this.slots[slot].getItemStackKey() == null || this.slots[slot].getItemStackKey() == toInsert) {
+            if (stack == null) {
+                stack = toInsert.getItemStack();
+            }
+            stack.setCount(amount);
```

Comment:
the remainder amount will not reflect te actual value if the overlay is reseted, but we can still use the item stack to avoid a copy. or we could hold them for a bit longer and use for future simulations hmmmm